aten/src/ATen/core/intrusive_ptr.h

```diff
@@ -0,0 +1,532 @@
+      : target_(rhs.target_) {
+    static_assert(
+        std::is_convertible<From*, TTarget*>::value,
+        "Type mismatch. intrusive_ptr copy constructor got pointer of wrong type.");
+    static_assert(
+        NullType::singleton() == FromNullType::singleton(),
+        "NullType mismatch. intrusive_ptr copy constructor got pointer with differing null value.");
+    retain();
+  }
+
+  ~intrusive_ptr() noexcept {
+    release();
+  }
+
+  intrusive_ptr& operator=(intrusive_ptr&& rhs) & noexcept {
+    return operator=<TTarget, NullType>(std::move(rhs));
+  }
+
+  template <class From, class FromNullType>
+      intrusive_ptr& operator=(intrusive_ptr<From, FromNullType>&& rhs) &
+      noexcept {
+    static_assert(
+        std::is_convertible<From*, TTarget*>::value,
+        "Type mismatch. intrusive_ptr move assignment got pointer of wrong type.");
+    static_assert(
+        NullType::singleton() == FromNullType::singleton(),
+        "NullType mismatch. intrusive_ptr move assignment got pointer with differing null value.");
+    release();
+    target_ = rhs.target_;
+    rhs.target_ = FromNullType::singleton();
+    return *this;
+  }
+
+  intrusive_ptr& operator=(const intrusive_ptr& rhs) & noexcept {
+    return operator=<TTarget, NullType>(rhs);
+  }
+
+  template <class From, class FromNullType>
+      intrusive_ptr& operator=(const intrusive_ptr<From, NullType>& rhs) &
+      noexcept {
+    static_assert(
+        std::is_convertible<From*, TTarget*>::value,
+        "Type mismatch. intrusive_ptr copy assignment got pointer of wrong type.");
+    static_assert(
+        NullType::singleton() == FromNullType::singleton(),
+        "NullType mismatch. intrusive_ptr copy assignment got pointer with differing null value.");
+    release();
+    target_ = rhs.target_;
+    retain();
+    return *this;
+  }
+
+  TTarget* get() const noexcept {
+    return target_;
+  }
+
+  const TTarget& operator*() const noexcept {
+    return *target_;
+  }
+
+  TTarget& operator*() noexcept {
+    return *target_;
+  }
+
+  const TTarget* operator->() const noexcept {
+    return target_;
+  }
+
+  TTarget* operator->() noexcept {
+    return target_;
+  }
+
+  void reset() noexcept {
+    release();
+  }
+
+  void swap(intrusive_ptr& rhs) noexcept {
+    TTarget* tmp = target_;
+    target_ = rhs.target_;
+    rhs.target_ = tmp;
+  }
+
+  // We do a lot of null-pointer checks in our code, good to have this be cheap.
+  bool defined() const noexcept {
+    return target_ != NullType::singleton();
+  }
+
+  size_t use_count() const noexcept {
+    if (target_ == NullType::singleton()) {
+      return 0;
+    }
+    return target_->refcount_.load();
+  }
+
+  bool unique() const noexcept {
+    return use_count() == 1;
+  }
+
+  template <class... Args>
+  static intrusive_ptr make(Args&&... args) {
+    auto result = intrusive_ptr(new TTarget(std::forward<Args>(args)...));
```

Comment:
This is a bit weird to me. Why doesn't the constructor for `TTarget` start off with `refcount_ == 1 && weakcount_ == 1`?

Reply:
Because I want to allow objects inheriting from `intrusive_ptr_target` to be instantiated on the stack if you don't need an `intrusive_ptr` to them, and then the refcount would semantically be zero. This is for example needed for the destructor assertion checking that refcount is zero on destruction.